Use zstd to compress state archives

jagerman opened the pull request:
These archives are huge (~85MB for the recent history, and 70MB for the archive), but are also extremely compressible.

This adds zstd dependency, which lets us compress them down to about 3.5MB for the archive (~20:1 compression), and 600kB (>100:1 compression), which allows us to significantly reduce the amount of I/O required during startup/shutdown (which always load/store), and during a subsystem rescan.

CPU time for compression here (on my system) is about 200ms for the long-term archive data, and around 40ms for the more compressible recent history.  Decompression of each adds around 20ms.

jagerman committed:
Use zstd to compress state archives

These archives are huge (~85MB for the recent history, and 70MB for the
archive), but are also extremely compressible.

This adds zstd dependency, which lets us compress them down to about
3.5MB for the archive (~20:1 compression), and 600kB (>100:1
compression), which allows us to significantly reduce the amount of I/O
required during startup/shutdown (which always load/store), and during a
subsystem rescan.

CPU time for compression here (on my system) is about 200ms for the
long-term archive data, and around 40ms for the more compressible recent
history.  Decompression of each adds around 20ms.

jagerman committed:
Add zstd to deps & update outdated requirements

jagerman committed:
static zstd: fix compilation issues

- cross-compilation was not working
- ccache was not being used
- add dep to deb snapshot builds

Doy-lee commented on src/cryptonote_core/service_node_list.cpp:
```diff
@@ -4941,6 +4953,65 @@ static std::string serialize_snl_directly(Archive& ar, service_node_list::state_
+        std::array<char, 4096> out_buf;
+        ZSTD_outBuffer output{.dst = out_buf.data(), .size = out_buf.size(), .pos = 0};
+
+        std::string decompressed;
```
You can query the decompressed size using `ZSTD_getFrameContentSize`

```cpp
uint64_t decompressed_size = ZSTD_getFrameContentSize(data.data(), data.size());
assert(decompressed_size != ZSTD_CONTENTSIZE_UNKNOWN && decompressed_size != ZSTD_CONTENTSIZE_ERROR);
```

Note that the decompressed size is always available if your use `ZSTD_compress` so in practice `ZSTD_CONTENSIZE_UNKNOWN` is not possible. `ZSTD_CONTENTSIZE_ERROR` can happen if the blob was corrupted so that one can be handled by returning `std::nullopt`.

jagerman replied:
Ah, thanks for that.  I've updated it with that to properly reserve now.

Doy-lee commented on src/cryptonote_core/service_node_list.cpp:
```diff
@@ -5036,20 +5107,29 @@ bool service_node_list::store(uint64_t state_height) {
             serialization::binary_string_archiver ar;
+
             std::string db_blob = serialize_db_blob(ar, archive_blob_list, nullptr);
+            db_blob = zstd_compress(
```
The first parameter in `zstd_compress` is `std::string_view` so the `std::move` can be removed.

jagerman replied:
Resolved.  The whole first version of `db_blob` here isn't needed (and was only like this because there was a debug line in between these two lines when I was working on this).

jagerman committed:
Extract & reserve decompressed size; simplify

- We can extract the decompressed size from the frame, to avoid multiple
  extra allocations during decompression.

- Inline variable being decompressed to simplify code.  (There was
  previously a debug line here in working code, but that is gone and the
  named-and-replaced temporary isn't needed).
